src/lib.rs

```diff
@@ -176,6 +230,106 @@ pub(crate) fn impl_ensure_dir_all(d: &openat::Dir, p: &Path, mode: libc::mode_t)
     Ok(())
 }
 
+/// A wrapper for atomically replacing a file.  The primary field
+/// to access here is the `writer`.  You can also configure the
+/// temporary prefix and suffix used for the temporary file before
+/// it is moved over the final destination.
+///
+/// To ensure that errors are handled correctly, you must explicitly
+/// invoke either `complete()` or `abandon()`.  Letting this value
+/// drop without invoking either of those will panic (unless the
+/// thread is already panicing, in which case this requirement is skipped).
+pub struct FileWriter<'a> {
+    /// Write to the destination file.
+    pub writer: std::io::BufWriter<std::fs::File>,
+    /// This string will be used as a prefix for the temporary file
+    pub tmp_prefix: String,
+    /// This string will be used as a suffix for the temporary file
+    pub tmp_suffix: String,
+
+    /// The target file name
+    destname: PathBuf,
+    /// The target directory
+    dir: &'a openat::Dir,
+    /// Used to ensure that either complete() or abandon() are invoked
+    bomb: drop_bomb::DropBomb,
+}
+
+impl<'a> FileWriter<'a> {
+    fn new(dir: &'a openat::Dir, f: std::fs::File, destname: PathBuf) -> Self {
+        Self {
+            writer: std::io::BufWriter::new(f),
+            destname,
+            tmp_prefix: ".tmp.".to_string(),
+            tmp_suffix: ".tmp".to_string(),
+            dir,
+            bomb: drop_bomb::DropBomb::new(
+                "FileWriter must be explicitly completed/abandoned to ensure errors are checked",
+            ),
+        }
+    }
+
+    /// Flush any outstanding buffered data and rename the temporary
+    /// file into place.  The provided closure will be invoked
+    /// on the real underlying file descriptor before it is
+    /// renamed into place.  You can use this to change file attributes.
+    /// For example, you can change the mode, extended attributes, or invoke
+    /// `fchmod()` to change ownership, etc.
+    pub fn complete_with<F>(mut self, f: F) -> io::Result<()>
+    where
+        F: Fn(&fs::File) -> io::Result<()>,
```

Comment:
`Fn(&mut impl std:io::Write)` perhaps?

Reply:
No, per the docs the caller must be able to access a real file descriptor to invoke `fchown/fchmod` etc.

Reply:
My bad, I totally missed the docstring and also misunderstood the intended usage of this.